sdk/src/signer/mod.rs

```diff
@@ -103,6 +109,25 @@ pub fn unique_signers(signers: Vec<&dyn Signer>) -> Vec<&dyn Signer> {
     signers.into_iter().unique_by(|s| s.pubkey()).collect()
 }
 
+/// The `EncodableKey` trait defines the interface by which cryptographic keys/keypairs are read,
+/// written, and derived from sources.
+pub trait EncodableKey: Sized {
+    fn pubkey_string(&self) -> Result<String, Box<dyn error::Error>>;
+    fn read_key<R: Read>(reader: &mut R) -> Result<Self, Box<dyn error::Error>>;
+    fn read_key_file<F: AsRef<Path>>(path: F) -> Result<Self, Box<dyn error::Error>>;
+    fn write_key<W: Write>(&self, writer: &mut W) -> Result<String, Box<dyn error::Error>>;
+    fn write_key_file<F: AsRef<Path>>(&self, outfile: F) -> Result<String, Box<dyn error::Error>>;
+    fn key_from_seed(seed: &[u8]) -> Result<Self, Box<dyn error::Error>>;
+    fn key_from_seed_and_derivation_path(
+        seed: &[u8],
+        derivation_path: Option<DerivationPath>,
```

Comment:
iirc, `DerivationPath` makes some assumptions about being bip32-ed25519/slip0011. does this complicate things?

Reply:
Yeah, ElGamal keys have the same cyclic group structure, so bip32 should still be applicable and same with authenticated encryption key for slip0011. But it would involve sophisticated parsers, so I was thinking about outputting a suitable error type for ElGamal and AE keys like [this](https://github.com/solana-labs/solana/pull/30901/files#diff-7d059813b85f149efcc5236cf7f8c29645b18bb9787cdaa05cec84711b06a8a6R309) (also see [above](https://github.com/solana-labs/solana/pull/30947#discussion_r1151326691)) Maybe we can support it in the future.